### DISCUSSION TITLE
Debugging full-stack moon applications in vscode?

### DISCUSSION THREAD

**Discussion by diraneyya** [moonrepo/moon#867] [Q&A]
Is there a recommended way to debug a full-stack application using `moon` in vscode? I tried to use a configuration of type `"node-terminal"` in `launch.json` but that does not seem to work smoothly. Namely, many debugger instances get attached, and inserting breakpoints from vscode does not work (breakpoints show as "unbound", not sure what this means).

I must admit that I do not understand debugging very well in the node ecosystem which frustrates me a little bit.

I tried the following configuration, but I am not happy with it either:
```json
    {
      "type": "node-terminal",
      "request": "launch",
      "name": "Debug Backend",
      "command": "moon run server:dev"
    }
    {
      "type": "chrome",
      "request": "launch",
      "name": "Debug Frontend",
      "url": "http://localhost:3000"
    },
```

When debugging the backend, my students need to click "Debug Backend" then launch the frontend manually using `moon run client:dev` from the terminal, but it gets messier than that.

Recently I have been experimenting with the possibilities in a monorepo by taking the Prisma client and the Prisma types outside of the server project (into `packages/prisma-client` and `packages/prisma-queries`, the latter of which has its own `:dev` local task which needs run simultaneously with `client:dev` and `server:dev`). So now when debugging the backend, the student needs to run `moon run client:dev queries:dev` after spinning the backend from the vscode debugger.

I am wondering if there is a better way to do any of this, since debugging is a very important part of software development. 

### Context

https://github.com/WildCodeSchool/2023-01-EN-Berlin-Remote2-Project3/commit/6a953fc5a1309bbea16fef7e22bd3411b939c929

### Vscode extension

It would be really nice if we can debug tasks from vscode using the moon extension. However the extension does not seem to provide a way to run a task across projects, for example, the equivalent of `moon run :dev`

**Comment by milesj** [comment]
@diraneyya Attaching the debugger to node processes that are ran via moon right now is... very complicated. For the node debugger to work, the `--inspect` or `--inspect-brk` options need to be passed to the `node` binary directly when executing.

Since moon abstracts this all away, it's not a simple thing. There's the `node.binExecArgs` setting (https://moonrepo.dev/docs/config/toolchain#binexecargs), but this would apply to _all_ node tasks, which is non-ideal.

At minimum, you would have to write this manually in the task:

```yaml
tasks:
  example:
    command: 'node --inspect ./some/file/to/run.js'
```

Long-term I want to support a `moon run --debugger` kind of system, but haven't had time to look into it.
